sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -361,6 +395,7 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     case (IntegerType, FloatType) => false
     case (LongType, FloatType) => false
     case (LongType, DoubleType) => false
+    case (TimestampType | TimestampNTZType, DateType) => true
```

Comment:
ditto

Reply:
have moved the logic to a new function: `needToCanonicalizeExpression`.